pkg/virt-controller/watch/vm.go

```diff
@@ -3121,6 +3121,12 @@ func (c *VMController) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachin
 				syncErr = &syncErrorImpl{fmt.Errorf("Error encountered when trying to check if VMI has interface with ordinal names (e.g.: eth1, eth2..): %v", err), HotPlugNetworkInterfaceErrorReason}
 			} else {
 				updatedVmiSpec := network.ApplyDynamicIfaceRequestOnVMI(vmCopy, vmiCopy, hasOrdinalIfaces)
+				// If a new interface is now present apply any device preferences we might have
+				if len(updatedVmiSpec.Domain.Devices.Interfaces) > len(vmiCopy.Spec.Domain.Devices.Interfaces) {
+					if _, err := c.applyDevicePreferences(vmCopy, vmiCopy); err != nil {
```

Comment:
It looks small and harmless, but I have a feeling it is a significant change compared to what existed so far.
If I understand correctly, before this addition, the VMI was applied the preferences only on creation.
Now, it is applied the preferences also on partial updates.

E.g. if the VM was updated while running, no preferences kicked in because the VMI was not updated.
However, if at some stage a new interface has been hot-plugged, this preference will be triggered, covering all changes so far to the VM.

Reply:
Preferences only provide preferred values for parts of the VirtualMachineInstanceSpec and do not overwrite anything already provided by a user within the VirtualMachine.

You are however correct that this starts to specifically apply the device related preferences during an interface hot plug on the VMI.